crates/service/src/middleware/attestation_signer.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2023-, Edge & Node, GraphOps, and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use alloy::primitives::Address;
+use axum::{
+    extract::{Request, State},
+    middleware::Next,
+    response::Response,
+};
+use indexer_attestation::AttestationSigner;
+use std::collections::HashMap;
+use tokio::sync::watch;
+
+use super::Allocation;
+
+#[derive(Clone)]
+pub struct AttestationState {
+    pub attestation_signers: watch::Receiver<HashMap<Address, AttestationSigner>>,
+}
+
+/// Injects the attestation signer to be used in the attestation
+///
+/// Needs Allocation Extension
+pub async fn signer_middleware(
+    State(state): State<AttestationState>,
+    mut request: Request,
+    next: Next,
+) -> Response {
+    if let Some(Allocation(allocation_id)) = request.extensions().get::<Allocation>() {
+        if let Some(signer) = state.attestation_signers.borrow().get(allocation_id) {
+            request.extensions_mut().insert(signer.clone());
+        }
+    }
+
+    next.run(request).await
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::middleware::{allocation::Allocation, signer_middleware, AttestationState};
+
+    use axum::{body::Body, http::Request, middleware::from_fn_with_state, routing::get, Router};
+    use indexer_attestation::AttestationSigner;
+    use indexer_monitor::attestation_signers;
+    use reqwest::StatusCode;
+    use test_assets::{DISPUTE_MANAGER_ADDRESS, INDEXER_ALLOCATIONS, INDEXER_MNEMONIC};
+    use tokio::sync::{mpsc::channel, watch};
+    use tower::Service;
+
+    #[tokio::test]
+    async fn test_attestation_signer_middleware() {
+        let allocations = (*INDEXER_ALLOCATIONS).clone();
+
+        let allocation = **allocations.keys().collect::<Vec<_>>().first().unwrap();
+
+        let (_, allocations_rx) = watch::channel(allocations);
+        let (_, dispute_manager_rx) = watch::channel(*DISPUTE_MANAGER_ADDRESS);
+        let attestation_signers = attestation_signers(
+            allocations_rx,
+            INDEXER_MNEMONIC.clone(),
+            1,
+            dispute_manager_rx,
+        );
+
+        let expected_signer = attestation_signers
+            .borrow()
+            .get(&allocation)
+            .unwrap()
+            .clone();
+
+        let state = AttestationState {
+            attestation_signers,
+        };
```

Comment:
From the beginning of the function until here, there is a setup process of , `state`, `allocation` and `expected_signer`. Maybe separate explicitly this setup like this?

Then, use the function below in the test function
```rust
    fn setup_attestation_state() -> (AttestationState, Address, AttestationSigner) {
        let allocations = (*INDEXER_ALLOCATIONS).clone();
        let allocation = **allocations.keys().collect::<Vec<_>>().first().unwrap();
        let (_, allocations_rx) = watch::channel(allocations);
        let (_, dispute_manager_rx) = watch::channel(*DISPUTE_MANAGER_ADDRESS);
        let attestation_signers = attestation_signers(
            allocations_rx,
            INDEXER_MNEMONIC.clone(),
            1,
            dispute_manager_rx,
        );

        let expected_signer = attestation_signers
            .borrow()
            .get(&allocation)
            .unwrap()
            .clone();

        let state = AttestationState {
            attestation_signers,
        };

        (state, allocation, expected_signer)
    }
    
    ```

Reply:
I usually do an extra function when it's reused across multiple tests. Could you explain why is this useful? Is it just for reading?

Reply:
Nevermind, it is just for reading in this case. I was under the impression this setup process was reused in the next tests multiple times too, but it is not the case. Marking as done.